### PR TITLE
Server Timing - Allow description and make duration optional

### LINKED ISSUE #879: Server-Timing: allow `desc` field, make `dur` optional
## Feature Description

The duration field is not mandatory for Server-Timing.

And description is very useful to provide more details.

You could actually also _only_ serve a description, e.g. `pageType;desc="index"`

IMHO this should be supported by this API to make it more flexible.

### PR DESCRIPTION
**Pull request by the-hercules**
## Summary

Fixes #879 

## Relevant technical choices
# Enhance Server-Timing API for MDN Specification Compliance

Enhances the Server-Timing API to be fully compliant with the MDN Server-Timing specification by making `duration` optional and adding support for a `desc` (description) field.

---

### `class-perflab-server-timing-metric.php`

- Added `$description` property to `Perflab_Server_Timing_Metric`
- Added `set_description()` and `get_description()` methods
- Used native PHP `string` type hint for `set_description()` instead of `string|mixed` with manual type checking
- Explicitly initialised `$description = null` for clarity

### `class-perflab-server-timing.php`

- Updated `format_metric_header_value()` to conditionally include `dur=` and `desc=` parameters
- Added sanitization for description values to prevent header injection:
  - Escapes `\` and `"`
  - Strips CR/LF control characters
  - Uses a combined `addcslashes()` call

---

## Tests

- Added `@covers` annotations to all new test methods
- Added tests for all four header output formats:
  - Name-only
  - Duration-only
  - Description-only
  - Combined duration + description
- Added edge-case tests for special characters in descriptions (`"`, `\`, `\n`, `\r`)
- Replaced vague `@phpstan-param array<string, mixed>` with:
  - `@param array<string, MetricArguments>`
  - Used `@phpstan-import-type` for accurate type definitions

---

## Supported Header Formats

| Format                  | Example Output                              |
|--------------------------|---------------------------------------------|
| Name only                | Server-Timing: missedCache              |
| Duration only            | `wp-metric;dur=45.23`                       |
| Description only         | `wp-metric;desc="Cache hit"`                |
| Duration + Description   | `wp-metric;dur=45.23;desc="Database queries"` |

---

## Testing

- All 137 Server-Timing tests passing on single site and multisite

---

## Fixes

Fixes #879

## Use of AI Tools

Github Copilot CLI was used to write tests. Reviewed the code afterwards.

### CODE REVIEW

File: plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing-metric.php [hunk 1]
```diff
@@ -75,4 +75,20 @@ public function test_measure_after_without_before(): void {
...
+	public function test_set_description_requires_string(): void {
+		$this->setExpectedIncorrectUsage( Perflab_Server_Timing_Metric::class . '::set_description' );
+
+		$this->metric->set_description( 123 );
+		$this->assertNull( $this->metric->get_description() );
+	}
+
+	public function test_get_description_returns_null_by_default(): void {
+		$this->assertNull( $this->metric->get_description() );
+	}
```
Comment (westonruter) [inline plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing-metric.php:93]: Please add `@covers` annotations for these methods.

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php [hunk 2]
```diff
@@ -142,4 +150,46 @@ public function measure_after(): void {
...
+		if ( ! is_string( $description ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( esc_html__( 'The %s parameter must be a string.', 'performance-lab' ), '$description' ),
+				''
+			);
+			return;
+		}
+
```
Comment (westonruter) [inline plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php]: Couldn't this just require a string?

```suggestion
	 * @param non-empty-string $description The metric description.
	 */
	public function set_description( string $description ): void {
```

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php [hunk 3]
```diff
@@ -37,6 +37,14 @@ class Perflab_Server_Timing_Metric {
 	 */
 	private $before_value;
 
+	/**
+	 * The metric description.
+	 *
+	 * @since n.e.x.t
+	 * @var string|null
+	 */
+	private $description;
```
Comment (westonruter) [inline plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php]:
```suggestion
	private $description = null;
```

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php [hunk 4]
```diff
@@ -290,23 +290,41 @@ function ( string $output, ?int $phase ): string {
...
+
+		if ( null !== $description ) {
+			// Sanitize description for HTTP header quoted-string format.
+			// Remove control characters (CR/LF) and escape backslashes and quotes.
+			$sanitized_description = preg_replace( '/[\r\n]/', '', $description );
+			if ( null === $sanitized_description ) {
+				$sanitized_description = '';
+			}
+			$sanitized_description = addcslashes( $sanitized_description, '\\' );
+			$sanitized_description = addcslashes( $sanitized_description, '"' );
```
Comment (westonruter) [inline plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php]: Can't these be combined?

```suggestion
			$sanitized_description = addcslashes( $sanitized_description, '\\"' 
```

File: plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php [hunk 5]
```diff
@@ -281,4 +281,143 @@ public function test_use_output_buffer( callable $set_up, bool $expected ): void
 		$set_up();
 		$this->assertSame( $expected, $this->server_timing->use_output_buffer() );
 	}
+
+	/**
+	 * @dataProvider data_get_header_with_description
+	 *
+	 * @phpstan-param array<string, mixed> $metrics
```
Comment (westonruter) [inline plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php]:
```suggestion
	 * @param array<string, mixed> $metrics
```

Comment (westonruter) [inline plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php]: Can we avoid `mixed`?

File: plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php [hunk 6]
```diff
@@ -281,4 +281,143 @@ public function test_use_output_buffer( callable $set_up, bool $expected ): void
...
+	 */
+	public function test_get_header_with_description( string $expected, array $metrics ): void {
+		foreach ( $metrics as $metric_slug => $args ) {
+			$this->server_timing->register_metric( $metric_slug, $args );
+		}
+		$this->assertSame( $expected, $this->server_timing->get_header() );
+	}
+
+	/**
+	 * @return array<string, mixed>
```
Comment (westonruter) [inline plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php]: Can we avoid `mixed`?

File: plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php [hunk 7]
```diff
@@ -281,4 +281,143 @@ public function test_use_output_buffer( callable $set_up, bool $expected ): void
...
+					),
+				),
+			),
+		);
+	}
+
+	/**
+	 * @dataProvider data_get_header_with_description_edge_cases
+	 *
+	 * @phpstan-param array<string, mixed> $metrics
```
Comment (westonruter) [inline plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php]:
```suggestion
	 * @param array<string, mixed> $metrics
```

ditto on `mixed`

File: plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php (hunk 5 above)
Comment (the-hercules) [reply plugins/performance-lab/tests/includes/server-timing/test-perflab-server-timing.php]: @westonruter Can I use

 `@param array<string, MetricArguments> $metrics`
after importing it
 `@phpstan-import-type MetricArguments from Perflab_Server_Timing` in places of mixed?

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php (hunk 2 above)
Comment (the-hercules) [reply plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php]: Hi @westonruter 

If we use a type hint without validation, PHP will coerce non-strings to strings (e.g., 123 → '123'). However, the test test_set_description_requires_string explicitly passes an integer and expects it to be rejected. This test now fails.

Questions:

1. Should we remove the validation and let PHP's type coercion handle it (meaning set_description(123) would silently become '123')?
2. Or should we keep the validation to reject non-strings explicitly?
3. In terms of API design, should description only accept strings, or can users pass numbers if they want?

Comment (westonruter) [reply plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php]: This is closely related to https://github.com/WordPress/performance/issues/2348. Compare:

* Without `declare( strict_types = 1 )`: https://3v4l.org/Mi8Qo
* With `declare( strict_types = 1 )`: https://3v4l.org/jEKNS

In the end, however, is it really a problem if someone passes a number here? Will someone even do that unexpectedly? Just by having the type set to `non-empty-string` for the `@param` and then having `string` PHP type hint for the param seems like it would be sufficient for informing developers they are expected to pass strings. It is in the name after all, that the argument is a _description_.

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php [hunk 8]
```diff
@@ -290,23 +290,40 @@ function ( string $output, ?int $phase ): string {
...
+			if ( is_float( $value ) ) {
+				$value = round( $value, 2 );
+			}
+			$parts[] = sprintf( 'dur=%s', $value );
+		}
+
+		if ( null !== $description ) {
+			// Sanitize description for HTTP header quoted-string format.
+			// Remove control characters (CR/LF) and escape backslashes and quotes.
+			$sanitized_description = preg_replace( '/[\r\n]/', '', $description );
```
Comment (westonruter) [inline plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php]: Using `str_replace()` is faster and simpler:

```suggestion
			$sanitized_description = str_replace( array( "\r", "\n" ), '', $description );
```

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php [hunk 9]
```diff
@@ -290,23 +290,40 @@ function ( string $output, ?int $phase ): string {
...
-		$value = $metric->get_value();
+		$value       = $metric->get_value();
+		$description = $metric->get_description();
 
-		// If no value is set, make sure it's just passed through.
-		if ( null === $value ) {
+		// If neither value nor description is set, skip this metric.
+		if ( null === $value && null === $description ) {
 			return null;
 		}
```
[the remark this answers is not in the document]
Comment (the-hercules) [reply plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php]: I'll update it to remove the "name only" format since that was inaccurate.Metric with neither duration nor description provides no useful information in the header. This PR only adds description as a new optional field alongside the existing optional duration.

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php [hunk 10]
```diff
@@ -142,4 +150,36 @@ public function measure_after(): void {
...
+	public function set_description( string $description ): void {
+		if ( 0 !== did_action( 'perflab_server_timing_send_header' ) && ! doing_action( 'perflab_server_timing_send_header' ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: WordPress action name */
+				sprintf( esc_html__( 'The method must be called before or during the %s action.', 'performance-lab' ), 'perflab_server_timing_send_header' ),
+				''
+			);
+			return;
+		}
```
[the remark this answers is not in the document]
Comment (the-hercules) [reply plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php:175]: The late-call guard in `set_description()` uses the exact same  
`did_action()` / `doing_action()` pattern as the existing `set_value()` method  
(lines 91–99).

Adding a test for this would require calling:

```php
do_action( 'perflab_server_timing_send_header' );
```

However, this triggers the global singleton to re-register default metrics
(e.g., before-template), which causes test isolation failures.

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php [hunk 11]
```diff
@@ -142,4 +150,36 @@ public function measure_after(): void {
...
+
+		$this->description = $description;
+	}
```
[the remark this answers is not in the document]
Comment (the-hercules) [reply plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php]: Leaving it as it is, relying on the type hint for the developer.

File: plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (westonruter) [reply plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php]: > Metric with neither duration nor description provides no useful information in the header.

Is this so? In the MDN docs, [this example](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Server-Timing#:~:text=//%20A%20metric%20with%20a%20name%20only) is provided:

```
// A metric with a name only
Server-Timing: missedCache
```

[the remark this answers is not in the document]
Comment (the-hercules) [reply plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php]: @westonruter Added support for name only.

**Comment by the-hercules** [comment]
@westonruter I’ve added support for name-only metrics as suggested. When you get a chance, could you please take a look?

**Review by westonruter — APPROVED** [review]
